### PR TITLE
Documented current protocol

### PR DESCRIPTION
**Pull request by janoszen**
This PR documents the status quo of the agent protocol and will serve as a basis for standardization.

### CODE REVIEW

File: docs.source/content/en/docs/How it works/_index.md [hunk 1]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
 []byte   L3OrJSON  | L3 or JSON
 ```
 
+## JSON messages
+
+JSON messages are used to configure the agent. There are 3 types of messages:
+
+- `request` are sent from the manager to the agent
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:28]: `the manager to the agent` -> `the manager to an agent` ?

File: docs.source/content/en/docs/How it works/_index.md [hunk 2]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
...
+```
+
+### The `request` body
+
+The request body has the following structure:
+
+```json
+{
+    "id": 1, //Unique ID for this request
+    "op": "operation key", //Currently the only operation supported is "configure"
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:48]: "operation key" -> "operator"

File: docs.source/content/en/docs/How it works/_index.md [hunk 3]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
...
+}
+```
+
+### The `configure` message arguments
+
+The `configure` message has the following arguments:
+
+```json
+{
+  "me": "1.2.3.4",
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:59]: IP is typically a local address like `192.168.42.100`, though 1.2.3.4 should work as well

File: docs.source/content/en/docs/How it works/_index.md [hunk 4]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
...
+  "forwards": [
+    // See Forward below
+  ],
+  "others": [
+    // See IPPortProto below
+  ],
+  "HostnameMap": {
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:66]: HostnameMap -> hostnameMap

File: docs.source/content/en/docs/How it works/_index.md [hunk 5]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
...
+    // See HTTP below
+  },
+  "socks": {
+    // See "Socks" below
+  },
+  "loopback": {
+    // See "Loopback" below
+  },
+  "statedir": {
+    "path": "/var/run/norouter...",
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:79]: statedir is typicalli like `/home/user/.norouter/agent`

File: docs.source/content/en/docs/How it works/_index.md [hunk 6]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
...
+}
+```
+
+### The `Forward` object
+
+The forward object has the following fields:
+
+```json
+{
+  "listen_port": 23
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:98]: Who uses 23/tcp (telnet) today? 😛

File: docs.source/content/en/docs/How it works/_index.md [hunk 7]
```diff
@@ -21,6 +21,201 @@ uint16be Reserved  | 0x0000
...
+      "loopback.disable",
+      //SOCKS
+      "socks",
+      //Creating ~/.norouter/agent/hostaliases file
+      "hostaliases",
+      //hostaliases using xip.io
+      "hostaliases.\"xip.io\"",
+      //Writing /etc/hosts when possible
+      "etchosts",
+      // Drwaing packets into a specific hosts. Only meaningful for HTTP and SOCKS proxy modes
```
Comment (AkihiroSuda) [inline docs.source/content/en/docs/How it works/_index.md:186]: Drwaing -> Drawing

**Comment by AkihiroSuda** [comment]
Thanks, could you sign off your commit? (`git commit -a -s --amend`)
https://github.com/norouter/norouter#contributing-to-norouter

File: docs.source/content/en/docs/How it works/_index.md (hunk 6 above)
Comment (janoszen) [reply docs.source/content/en/docs/How it works/_index.md:98]: You'd be surprised. :D Anyway, it was meant as a joke. :D

Comment (janoszen) [reply docs.source/content/en/docs/How it works/_index.md:98]: Do you want me to change it?

Comment (AkihiroSuda) [reply docs.source/content/en/docs/How it works/_index.md:98]: Either is fine :+1:

**Comment by janoszen** [comment]
Changes made and signed off. (Sorry, I thought a GPG signature would be enough.)

**Review by AkihiroSuda — APPROVED** [review]
Thanks!

**Pull request merged**